File: python/tree/LC_94_BinaryTreeInorderTraversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/
from typing import Optional, List
from tree.treeNode import TreeNode


class Solution:
    def traverse_inorder(self, vals, node):
        if node:
            self.traverse_inorder(vals, node.left)
            vals.append(node.val)
            self.traverse_inorder(vals, node.right)

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        vals = []
        self.traverse_inorder(vals, root)
        return vals

File: java/src/tree/LC_94_BinaryTreeInorderTraversal.java
package tree;

import java.util.ArrayList;
import java.util.List;

public class LC_94_BinaryTreeInorderTraversal {
    public void traverseInorder(List<Integer> list, TreeNode node) {
        if (node != null) {
            traverseInorder(list, node.left);
            list.add(node.val);
            traverseInorder(list, node.right);
        }
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        traverseInorder(list, root);
        return list;
    }

    public static void main(String[] args) {
        TreeNode root = TreeNode.buildTree(1,2,3,4,5,null,8,null,null,6,7,9);
        LC_94_BinaryTreeInorderTraversal traversel = new LC_94_BinaryTreeInorderTraversal();
        System.out.println(traversel.inorderTraversal(root));
    }

}
